lib/smack/SmackRep.cpp

```diff
@@ -758,36 +935,99 @@ const Expr* SmackRep::declareIsExternal(const Expr* e) {
 }
 
 string SmackRep::getPtrType() {
-  return "int";
+  if (SmackOptions::BitVectors)
+    return "ref";
+  else
+    return "int";
 }
 
 string SmackRep::memcpyProc(int dstReg, int srcReg) {
   stringstream s;
 
   if (SmackOptions::MemoryModelImpls) {
-    s << "procedure $memcpy." << dstReg << "." << srcReg;
-    s << "(dest: int, src: int, len: int, align: int, isvolatile: bool)" << endl;
-    s << "modifies " << memReg(dstReg) << ";" << endl;
-    s << "{" << endl;
-    s << "  var $oldSrc: [" << getPtrType() << "] " << getPtrType() << ";" << endl;
-    s << "  var $oldDst: [" << getPtrType() << "] " << getPtrType() << ";" << endl;
-    s << "  $oldSrc := " << memReg(srcReg) << ";" << endl;
-    s << "  $oldDst := " << memReg(dstReg) << ";" << endl;
-    s << "  havoc " << memReg(dstReg) << ";" << endl;
-    s << "  assume (forall x:int :: dest <= x && x < dest + len ==> "
-      << memReg(dstReg) << "[x] == $oldSrc[src - dest + x]);" << endl;
-    s << "  assume (forall x:int :: !(dest <= x && x < dest + len) ==> "
-      << memReg(dstReg) << "[x] == $oldDst[x]);" << endl;
-    s << "}" << endl;
+    if (SmackOptions::BitVectors) {
+      // TODO
+      s << "procedure $memcpy." << dstReg << "." << srcReg;
+      s << "(dest: ref, src: ref, len: ref, align: ref, isvolatile: bool)" << endl;
```

Comment:
Well either way, they should not be forced to be `ref` types, should they? Perhaps you will need a many declarations of `$memcpy`, one for each variation to the argument types.